storage-node/src/storage_manager.rs

```diff
@@ -272,4 +275,95 @@ mod tests {
         );
         assert!(delta_time_hit_disk > delta_time_hit_mem);
     }
+
+    #[tokio::test]
+    async fn test_storage_manager_parallel_1() {
+        let disk_cache = LruReplacer::new(1000000);
+
+        let tmp = tempfile::tempdir().unwrap();
+        let disk_cache_base_path = tmp.path().to_owned();
+
+        let data_store_cache = MemDiskStoreCache::new(
+            disk_cache,
+            disk_cache_base_path.display().to_string(),
+            None,
+            None,
+        );
+        let storage_manager = StorageManager::new(data_store_cache);
+
+        let request_path_bucket1 = "tests-parquet".to_string();
+        let request_path_keys1 = vec!["userdata1.parquet".to_string()];
+        let request_data1 = RequestParams::MockS3((request_path_bucket1, request_path_keys1));
+
+        let request_path_bucket2 = "tests-parquet".to_string();
+        let request_path_keys2 = vec!["userdata2.parquet".to_string()];
+        let request_data2 = RequestParams::MockS3((request_path_bucket2, request_path_keys2));
+
+        let get_data_fut_1 = storage_manager.get_data(request_data1.clone(), true);
+        let get_data_fut_2 = storage_manager.get_data(request_data1.clone(), true);
+        let get_data_fut_3 = storage_manager.get_data(request_data2.clone(), true);
+        let get_data_fut_4 = storage_manager.get_data(request_data1.clone(), true);
+        let result = join!(
+            get_data_fut_1,
+            get_data_fut_2,
+            get_data_fut_3,
+            get_data_fut_4
+        );
+        assert!(result.0.is_ok());
+        assert_eq!(consume_receiver(result.0.unwrap()).await, 113629);
+        assert!(result.1.is_ok());
+        assert_eq!(consume_receiver(result.1.unwrap()).await, 113629);
+        assert!(result.2.is_ok());
+        assert_eq!(consume_receiver(result.2.unwrap()).await, 112193);
+        assert!(result.3.is_ok());
+        assert_eq!(consume_receiver(result.3.unwrap()).await, 113629);
```

Comment:
Are the futures guaranteed to return in this order?

Reply:
Yes!